File: ai-service/services/safety/clinical_conflict.py
"""
Clinical contradiction detection across authors.

The organisers' hint: "Conflict detection. Scope this for allergies,
medications, dosages, etc because human-human contradictions do exist between
clinician and nurse and staff notes."

This is a different problem from services/conflict.py. That module resolves
competing EDITS to the same section — one wins, deterministically. This module
detects competing CLAIMS about the same clinical fact, and deliberately does not
resolve them.

The distinction matters. If Dr A wrote 50mg and Nurse B wrote 500mg, the system
has no basis to decide which is correct, and picking one would manufacture false
certainty about a dosing decision. So it surfaces the delta with both verbatim
quotes and their authors, and asks a human. Silence would be worse; a guess
would be dangerous.

Detection is deterministic regex over the stored text — no model involved — so
it cannot hallucinate a contradiction that is not there, and it produces the
same answer on every run.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Iterable


class ConflictClass(str, Enum):
    """Contradiction categories, ordered by how badly they can hurt someone."""

    ALLERGY = "allergy"          # highest: an allergy contradiction can kill
    DOSAGE = "dosage"
    MEDICATION = "medication"
    VITAL = "vital"

    @property
    def severity(self) -> str:
        return "critical" if self is ConflictClass.ALLERGY else "high"


_UNIT = r"(?:mg|mcg|µg|g|ml|units?|iu)"

# Dose statements: "Lisinopril 10mg", "Metformin 1000 mg BID".
_DOSE = re.compile(
    rf"\b([A-Z][a-z]{{3,}}(?:ol|in|ide|ine|pril|artan|statin|azole|mycin|cillin|formin|pam|done))\b"
    rf"\s*(?:at|to|-)?\s*(\d+(?:\.\d+)?)\s*({_UNIT})\b",
    re.IGNORECASE,
)

# Positive allergy assertions.
_ALLERGY = re.compile(
    r"\ballerg(?:y|ic|ies)\s+(?:to\s+)?([A-Za-z][A-Za-z\- ]{2,30}?)\b(?=[.,;)]|\s+(?:and|with|but)\b|$)",
    re.IGNORECASE,
)

# Explicit denials — "not allergic to penicillin", "no known allergies".
_NO_ALLERGY = re.compile(
    r"\b(?:no\s+known\s+(?:drug\s+)?allergies|not\s+allergic\s+to\s+([A-Za-z][A-Za-z\- ]{2,30}?)\b)",
    re.IGNORECASE,
)

_VITALS = {
    "blood_pressure": re.compile(r"\bBP\s*:?\s*(\d{2,3}\s*/\s*\d{2,3})", re.IGNORECASE),
    "heart_rate": re.compile(r"\bHR\s*:?\s*(\d{2,3})\b", re.IGNORECASE),
    "a1c": re.compile(r"\b(?:A1C|HbA1c)\s*:?\s*(\d+(?:\.\d+)?)\s*%?", re.IGNORECASE),
}


@dataclass
class Assertion:
    """One clinical claim, tied to who said it and the exact words they used."""

    entity: str
    value: str
    conflict_class: ConflictClass
    author_id: str | None
    author_role: str
    entry_id: str
    quote: str
    timestamp: Any = None

    @property
    def normalized_value(self) -> str:
        return re.sub(r"\s+", "", self.value.lower())


@dataclass
class ClinicalConflict:
    """A contradiction, presented for human resolution — never auto-resolved."""

    entity: str
    conflict_class: ConflictClass
    assertions: list[Assertion] = field(default_factory=list)

    @property
    def severity(self) -> str:
        return self.conflict_class.severity

    @property
    def requires_human_resolution(self) -> bool:
        """Always true. The system reports the delta; a clinician decides."""
        return True

    @property
    def distinct_values(self) -> list[str]:
        seen: list[str] = []
        for a in self.assertions:
            if a.value not in seen:
                seen.append(a.value)
        return seen

    def describe(self) -> str:
        parts = [
            f"{a.author_role} noted {a.value!r}" for a in self.assertions
        ]
        return f"{self.conflict_class.value} conflict on {self.entity}: " + " vs ".join(parts)

    def to_metadata(self) -> dict[str, Any]:
        """
        Audit payload. Carries the quotes on purpose — a clinician resolving a
        dosing contradiction must see the exact wording of both claims. This is
        internal-only and never enters the patient-facing path.
        """
        return {
            "conflict_class": self.conflict_class.value,
            "entity": self.entity,
            "severity": self.severity,
            "requires_human_resolution": True,
            "claims": [
                {
                    "author_role": a.author_role,
                    "author_id": a.author_id,
                    "entry_id": a.entry_id,
                    "value": a.value,
                    "quote": a.quote,
                    "timestamp": a.timestamp,
                }
                for a in self.assertions
            ],
        }


def _quote_around(text: str, start: int, end: int, pad: int = 60) -> str:
    """The surrounding sentence fragment, so the claim is readable in context."""
    return text[max(0, start - pad) : min(len(text), end + pad)].strip()


def extract_assertions(entry: dict[str, Any]) -> list[Assertion]:
    """Pull medication, allergy and vital claims out of one timeline entry."""
    text = entry.get("content_text") or ""
    if not text:
        return []

    common = {
        "author_id": entry.get("author_id"),
        "author_role": entry.get("author_role", "unknown"),
        "entry_id": entry.get("id", ""),
        "timestamp": entry.get("created_at"),
    }
    found: list[Assertion] = []

    for m in _DOSE.finditer(text):
        drug, amount, unit = m.group(1), m.group(2), m.group(3)
        found.append(Assertion(
            entity=drug.lower(), value=f"{amount}{unit.lower()}",
            conflict_class=ConflictClass.DOSAGE,
            quote=_quote_around(text, m.start(), m.end()), **common,
        ))

    for m in _NO_ALLERGY.finditer(text):
        allergen = (m.group(1) or "any").strip().lower()
        found.append(Assertion(
            entity=f"allergy:{allergen}", value="none",
            conflict_class=ConflictClass.ALLERGY,
            quote=_quote_around(text, m.start(), m.end()), **common,
        ))

    denied_spans = [(m.start(), m.end()) for m in _NO_ALLERGY.finditer(text)]
    for m in _ALLERGY.finditer(text):
        # Skip a positive match that sits inside an explicit denial.
        if any(s <= m.start() < e for s, e in denied_spans):
            continue
        allergen = m.group(1).strip().lower()
        found.append(Assertion(
            entity=f"allergy:{allergen}", value="present",
            conflict_class=ConflictClass.ALLERGY,
            quote=_quote_around(text, m.start(), m.end()), **common,
        ))

    for name, pattern in _VITALS.items():
        for m in pattern.finditer(text):
            found.append(Assertion(
                entity=name, value=m.group(1).strip(),
                conflict_class=ConflictClass.VITAL,
                quote=_quote_around(text, m.start(), m.end()), **common,
            ))

    return found


def detect_conflicts(
    entries: Iterable[dict[str, Any]],
    *,
    include_same_author: bool = False,
) -> list[ClinicalConflict]:
    """
    Find entities asserted with different values across entries.

    By default only cross-author disagreements are reported. One author
    revising their own note over time is a correction, not a contradiction, and
    reporting it would be noise — which is exactly how alert fatigue starts.

    Vitals are excluded from conflict reporting entirely: a blood pressure
    differing between April and October is the timeline working as intended, not
    a disagreement.
    """
    assertions: list[Assertion] = []
    for entry in entries:
        assertions.extend(extract_assertions(entry))

    grouped: dict[tuple[str, ConflictClass], list[Assertion]] = {}
    for a in assertions:
        if a.conflict_class is ConflictClass.VITAL:
            continue  # measurements change legitimately over time
        grouped.setdefault((a.entity, a.conflict_class), []).append(a)

    conflicts: list[ClinicalConflict] = []
    for (entity, cls), group in grouped.items():
        if len({a.normalized_value for a in group}) < 2:
            continue
        if not include_same_author and len({a.author_id for a in group}) < 2:
            continue
        conflicts.append(ClinicalConflict(
            entity=entity, conflict_class=cls,
            assertions=sorted(group, key=lambda a: str(a.timestamp or "")),
        ))

    # Allergy contradictions first — they are the ones that kill people.
    order = {ConflictClass.ALLERGY: 0, ConflictClass.DOSAGE: 1,
             ConflictClass.MEDICATION: 2, ConflictClass.VITAL: 3}
    conflicts.sort(key=lambda c: (order[c.conflict_class], c.entity))
    return conflicts

File: frontend/lib/safety/clinical-conflict.ts
import type { ClinicalConflict, ConflictClaim, TimelineEntry } from '@/lib/types';

/**
 * Client-side clinical contradiction detection.
 *
 * A TypeScript port of `ai-service/services/safety/clinical_conflict.py`, kept
 * deliberately in lockstep with it — same patterns, same exclusions, same
 * ordering. The Python module is the tested reference (see
 * `tests/test_clinical_safety.py::TestClinicalConflictDetection`); this exists
 * so the UI can flag a contradiction the moment a timeline loads, without a
 * round trip to the AI service.
 *
 * Deterministic regex, no model. It cannot hallucinate a contradiction that is
 * not in the text, and it produces the same answer every run.
 *
 * The system surfaces the delta and never arbitrates. If a clinician wrote 10mg
 * and a nurse recorded 100mg, there is no basis to decide which is right, and
 * choosing would manufacture false certainty about a dosing decision.
 */

const UNIT = '(?:mg|mcg|µg|g|ml|units?|iu)';

const DOSE = new RegExp(
  `\\b([A-Z][a-z]{3,}(?:ol|in|ide|ine|pril|artan|statin|azole|mycin|cillin|formin|pam|done))\\b` +
    `\\s*(?:at|to|-)?\\s*(\\d+(?:\\.\\d+)?)\\s*(${UNIT})\\b`,
  'gi',
);

/**
 * Dosages written as words.
 *
 * Kept in lockstep with `words_to_number` in the Python reference. Comparing
 * "one hundred mg" against "10mg" as raw strings finds no conflict, so words
 * normalise to digits before any comparison. A contradiction expressed in prose
 * is still a contradiction — arguably a more dangerous one, because it reads as
 * deliberate rather than as a typo.
 */
const NUMBER_WORDS: Record<string, number> = {
  zero: 0, one: 1, two: 2, three: 3, four: 4, five: 5, six: 6, seven: 7,
  eight: 8, nine: 9, ten: 10, eleven: 11, twelve: 12, thirteen: 13,
  fourteen: 14, fifteen: 15, sixteen: 16, seventeen: 17, eighteen: 18,
  nineteen: 19, twenty: 20, thirty: 30, forty: 40, fifty: 50, sixty: 60,
  seventy: 70, eighty: 80, ninety: 90,
};
const SCALE_WORDS: Record<string, number> = { hundred: 100, thousand: 1000 };

const ALL_NUMBER_WORDS = [...Object.keys(NUMBER_WORDS), ...Object.keys(SCALE_WORDS)]
  .sort((a, b) => b.length - a.length)
  .join('|');

export function wordsToNumber(phrase: string): number | null {
  let total = 0;
  let current = 0;
  let seen = false;
  for (const token of phrase.toLowerCase().match(/[a-z]+/g) ?? []) {
    if (token in NUMBER_WORDS) {
      current += NUMBER_WORDS[token];
      seen = true;
    } else if (token in SCALE_WORDS) {
      const scale = SCALE_WORDS[token];
      current = (current || 1) * scale;
      if (scale >= 1000) {
        total += current;
        current = 0;
      }
      seen = true;
    } else if (token !== 'and') {
      return null;
    }
  }
  return seen ? total + current : null;
}

const DOSE_WORDS = new RegExp(
  `\\b([A-Z][a-z]{3,}(?:ol|in|ide|ine|pril|artan|statin|azole|mycin|cillin|formin|pam|done))\\b` +
    `\\s*(?:at|to|-)?\\s*((?:${ALL_NUMBER_WORDS})(?:[\\s-]+(?:and[\\s-]+)?(?:${ALL_NUMBER_WORDS}))*)` +
    `\\s*(${UNIT})\\b`,
  'gi',
);

const ALLERGY =
  /\ballerg(?:y|ic|ies)\s+(?:to\s+)?([A-Za-z][A-Za-z\- ]{2,30}?)\b(?=[.,;)]|\s+(?:and|with|but)\b|$)/gi;

const NO_ALLERGY =
  /\b(?:no\s+known\s+(?:drug\s+)?allergies|not\s+allergic\s+to\s+([A-Za-z][A-Za-z\- ]{2,30}?)\b)/gi;

interface Assertion extends ConflictClaim {
  entity: string;
  conflictClass: ClinicalConflict['conflict_class'];
}

function quoteAround(text: string, start: number, end: number, pad = 60): string {
  return text.slice(Math.max(0, start - pad), Math.min(text.length, end + pad)).trim();
}

function extractAssertions(entry: TimelineEntry): Assertion[] {
  const text = entry.content_text ?? '';
  if (!text) return [];

  const base = {
    author_role: entry.author_role ?? 'unknown',
    author_id: entry.author_id ?? null,
    entry_id: entry.id,
    timestamp: entry.created_at,
  };
  const found: Assertion[] = [];

  for (const m of text.matchAll(DOSE)) {
    found.push({
      ...base,
      entity: m[1].toLowerCase(),
      value: `${m[2]}${m[3].toLowerCase()}`,
      conflictClass: 'dosage',
      quote: quoteAround(text, m.index ?? 0, (m.index ?? 0) + m[0].length),
    });
  }

  for (const m of text.matchAll(DOSE_WORDS)) {
    const value = wordsToNumber(m[2]);
    if (value === null) continue;
    const normalised = `${value}${m[3].toLowerCase()}`;
    const start = m.index ?? 0;
    if (found.some((a) => a.entity === m[1].toLowerCase() && a.value === normalised)) continue;
    found.push({
      ...base,
      entity: m[1].toLowerCase(),
      value: normalised,
      conflictClass: 'dosage',
      quote: quoteAround(text, start, start + m[0].length),
    });
  }

  const deniedSpans: Array<[number, number]> = [];
  for (const m of text.matchAll(NO_ALLERGY)) {
    const start = m.index ?? 0;
    deniedSpans.push([start, start + m[0].length]);
    found.push({
      ...base,
      entity: `allergy:${(m[1] ?? 'any').trim().toLowerCase()}`,
      value: 'none',
      conflictClass: 'allergy',
      quote: quoteAround(text, start, start + m[0].length),
    });
  }

  for (const m of text.matchAll(ALLERGY)) {
    const start = m.index ?? 0;
    // Skip a positive match sitting inside an explicit denial.
    if (deniedSpans.some(([s, e]) => s <= start && start < e)) continue;
    found.push({
      ...base,
      entity: `allergy:${m[1].trim().toLowerCase()}`,
      value: 'present',
      conflictClass: 'allergy',
      quote: quoteAround(text, start, start + m[0].length),
    });
  }

  return found;
}

const CLASS_ORDER: Record<ClinicalConflict['conflict_class'], number> = {
  allergy: 0,
  dosage: 1,
  medication: 2,
  vital: 3,
};

export function detectClinicalConflicts(entries: TimelineEntry[]): ClinicalConflict[] {
  const assertions = entries.flatMap(extractAssertions);

  const grouped = new Map<string, Assertion[]>();
  for (const a of assertions) {
    const key = `${a.entity}|${a.conflictClass}`;
    grouped.set(key, [...(grouped.get(key) ?? []), a]);
  }

  const conflicts: ClinicalConflict[] = [];
  for (const group of grouped.values()) {
    const values = new Set(group.map((a) => a.value.toLowerCase().replace(/\s+/g, '')));
    // Same-author revision over time is a correction, not a disagreement.
    const authors = new Set(group.map((a) => a.author_id));
    if (values.size < 2 || authors.size < 2) continue;

    conflicts.push({
      conflict_class: group[0].conflictClass,
      entity: group[0].entity,
      severity: group[0].conflictClass === 'allergy' ? 'critical' : 'high',
      requires_human_resolution: true,
      claims: [...group]
        .sort((a, b) => String(a.timestamp ?? '').localeCompare(String(b.timestamp ?? '')))
        .map(({ author_role, author_id, entry_id, value, quote, timestamp }) => ({
          author_role,
          author_id,
          entry_id,
          value,
          quote,
          timestamp,
        })),
    });
  }

  // Allergy contradictions first — they are the ones that kill people.
  return conflicts.sort(
    (a, b) =>
      CLASS_ORDER[a.conflict_class] - CLASS_ORDER[b.conflict_class] ||
      a.entity.localeCompare(b.entity),
  );
}
